### PR TITLE
Se 428 squash rule storage and processor diamonds

### PR DESCRIPTION
**Pull request by TJ-Everett**

### CODE REVIEW

File: docs/userGuides/deployment/CREATE-APP-RULES.md [hunk 1]
```diff
@@ -6,15 +6,15 @@
 ## Application Rule Implementation Overview
 1.  Ensure the [environment variables][environment-url] are set correctly.
 2.  Create an application rule:
-    -  Get the rule creation function from the [Application Rule Directory][appRuleDirectory-url] and invoke it on the RuleStorageDiamond, sending in the required parameters. NOTE: Each rule requires a different parameter set. For local deployments, the RuleStorageDiamond address can be found in previous steps, otherwise consult the [Deployment Directory][deploymentDirectory-url]. 
+    -  Get the rule creation function from the [Application Rule Directory][appRuleDirectory-url] and invoke it on the RuleProcssorDiamond, sending in the required parameters. NOTE: Each rule requires a different parameter set. For local deployments, the RuleProcssorDiamond address can be found in previous steps, otherwise consult the [Deployment Directory][deploymentDirectory-url].
```
Comment (ShaneDuncan602) [inline docs/userGuides/deployment/CREATE-APP-RULES.md]: Looks like a copy/paste error on RuleProcssorDiamond

File: script/DeployProcessorDiamond.s.sol [hunk 2]
```diff
@@ -77,7 +71,11 @@ contract DeployAllModulesScript is Script {
             "ERC20TaggedRuleProcessorFacet",
             "ERC721TaggedRuleProcessorFacet",
             "RiskTaggedRuleProcessorFacet",
-            "RuleApplicationValidationFacet"
+            "RuleApplicationValidationFacet",
+            "RuleDataFacet",
+            "TaggedRuleDataFacet",
+            "AppRuleDataFacet",
+            "FeeRuleDataFacet"
```
Comment (ShaneDuncan602) [inline script/DeployProcessorDiamond.s.sol:78]: I hope this all works in a single script. It's a a lot of deployments.

File: src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
-import {AppRuleDataFacet} from "../ruleStorage/AppRuleDataFacet.sol";
-import {IApplicationRules as Application} from "../ruleStorage/RuleDataInterfaces.sol";
-import {IRuleProcessorErrors, IAccessLevelErrors} from "../../interfaces/IErrors.sol";
+import {RuleProcessorDiamondLib as processorDiamond, RuleDataStorage} from "./RuleProcessorDiamondLib.sol";
+import {AppRuleDataFacet} from "./AppRuleDataFacet.sol";
+import {RuleStoragePositionLib as Storage} from "./RuleStoragePositionLib.sol";
+import {IRuleStorage as RuleS} from "./IRuleStorage.sol";
+import {IApplicationRules as Application} from "./RuleDataInterfaces.sol";
+import {IInputErrors, IRuleProcessorErrors, IAccessLevelErrors} from "../../interfaces/IErrors.sol";
+import "./RuleProcessorCommonLib.sol";
```
Comment (ShaneDuncan602) [inline src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol]: I recommend we put the common imports into an abstract contract that can be inherited by all the processor facets. We have a lot of repetition here. RuleProcessorDiamondLib, RuleStoragePositionLib, IRuleStorage, RuleDataInterfaces, IErrors, and RuleProcessorCommonLib are shared by all processor facets.

File: src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol [hunk 4]
```diff
@@ -22,15 +29,35 @@ contract ApplicationAccessLevelProcessorFacet is IRuleProcessorErrors, IAccessLe
      * @param _amountToTransfer total USD amount to be transferred with 18 decimals of precision
      */
     function checkAccBalanceByAccessLevel(uint32 _ruleId, uint8 _accessLevel, uint128 _balance, uint128 _amountToTransfer) external view {
-        AppRuleDataFacet data = AppRuleDataFacet(actionDiamond.ruleDataStorage().rules);
         /// Get the account's AccessLevel
-        if (data.getTotalAccessLevelBalanceRules() != 0) {
-            uint48 max = data.getAccessLevelBalanceRule(_ruleId, _accessLevel);
+        if (getTotalAccessLevelBalanceRules() != 0) {
```
Comment (ShaneDuncan602) [inline src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol]: We don't need these getTotal function calls here. The rule getter takes care of the scenario where the rule doesn't exist. We can remove the getTotal from all of these. (The call, not the function itself.)

**Review by ShaneDuncan602 — CHANGES_REQUESTED** [review]
This is very good and extremely thorough. I just have a couple of suggestions and found a typo. Great job.

File: src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol (hunk 3 above)
Comment (TJ-Everett) [reply src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol]: Love this idea and working on it now.

**Review by ShaneDuncan602 — COMMENTED** [review]
Also, I found more references to the Rule Storage Diamond in the deployment docs. Check out the deployment folder.

File: src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol (hunk 4 above)
Comment (TJ-Everett) [reply src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol]: That is a great point - thanks for catching this. 
I have some rough gas numbers for before this and I will rerun once I remove these and let ya know how much efficiency we gain just from this.

**Comment by TJ-Everett** [comment]
> Also, I found more references to the Rule Storage Diamond in the deployment docs. Check out the deployment folder.

I will double check - some of those (if using grep) are from previously deployed addresses [here](https://github.com/thrackle-io/tron/blob/external/docs/userGuides/deployment/DEPLOYMENT-DIRECTORY.md). I will make sure those are the only ones.

File: docs/userGuides/GLOSSARY.md [hunk 5]
```diff
@@ -16,8 +16,8 @@
...
 | Access-Tier Provider      | An external provider that rates or segments users based on external criteria such as Know Your Customer (KYC) solutions.                                     |
 | Permission Roles          | Roles used by AppManager. They include: Admin, Access Tier Admin, and Risk Admin.                                                                            |
-| Application Rule          | Rule applied to all protocol supported assets. They are created using the protocol's RuleStorageDiamond and applied in the application's AppManager.        |
-| Token Specific Rule       | Rule applied to a specific protocol supported entity. They are created using the protocol's RuleStorageDiamond and applied in the token's Handler.        |
+| Application Rule          | Rule applied to all protocol supported assets. They are created using the protocol's RuleProcssorDiamond and applied in the application's AppManager.        |
+| Token Specific Rule       | Rule applied to a specific protocol supported entity. They are created using the protocol's RuleProcssorDiamond and applied in the token's Handler.        |
```
Comment (oscarsernarosero) [inline docs/userGuides/GLOSSARY.md]: RuleProcessor*

File: docs/userGuides/deployment/DEPLOY-PROTOCOL.md [hunk 6]
```diff
@@ -45,14 +45,16 @@ is an overview of this deployment process:
 ```
       "hash": "0x1902f5f3c6f2ed24ae3a64c8ddb41e72fb71b57c3404278c965dee920aa6f40f",
       "transactionType": "CREATE",
-      "contractName": "RuleStorageDiamond",
+      "contractName": "RuleProcssorDiamond",
```
Comment (oscarsernarosero) [inline docs/userGuides/deployment/DEPLOY-PROTOCOL.md]: I think Processor is misspelled in general, even in the contract.

File: script/DeployProcessorDiamond.s.sol (hunk 2 above)
Comment (oscarsernarosero) [reply script/DeployProcessorDiamond.s.sol:78]: Hm. I second that

File: src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol [hunk 7]
```diff
@@ -22,13 +23,31 @@ contract ApplicationAccessLevelProcessorFacet is IRuleProcessorErrors, IAccessLe
...
+        return data.accessRulesPerToken[_index][_accessLevel];
+    }
+
+    /**
+     * @dev Function to get total AccessLevel Balance rules
+     * @return Total length of array
+     */
+    function getTotalAccessLevelBalanceRules() public view returns (uint32) {
+        RuleS.AccessLevelRuleS storage data = Storage.accessStorage();
+        return data.accessRuleIndex;
```
Comment (oscarsernarosero) [inline src/economic/ruleProcessor/ApplicationAccessLevelProcessorFacet.sol:50]: very nice Tayler

**Review by oscarsernarosero — APPROVED** [review]
This was really good Tayler. I love that you relocated the getRule and getTotalRules from the storage facets to the processor facets since this makes the whole protocol way more efficient. I just noticed a propagated typo, but Im approving anyways since I know you'll fix it before the merge.

**Review by ShaneDuncan602 — APPROVED** [review]
Looks fantastic

**Pull request merged**